store/secretsmanagerstore.go

```diff
@@ -0,0 +1,411 @@
+package store
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+)
+
+// secretValueObject is the serialized format for storing secrets
+// as a SecretsManager SecretValue
+type secretValueObject map[string]secretValueVersion
+
+// secretValueVersion holds all the metadata for a specific version
+// of a secret
+type secretValueVersion struct {
+	Created   time.Time `json:"created"`
+	CreatedBy string    `json:"created_by"`
+	Version   int       `json:"version"`
+	Value     string    `json:"value"`
+}
+
+// ensure SecretsManagerStore confirms to Store interface
+var _ Store = &SecretsManagerStore{}
+
+// SecretsManagerStore implements the Store interface for storing secrets in SSM Parameter
+// Store
+type SecretsManagerStore struct {
+	svc    secretsmanageriface.SecretsManagerAPI
+	stsSvc stsiface.STSAPI
+}
+
+// NewSecretsManagerStore creates a new SecretsManagerStore
+func NewSecretsManagerStore(numRetries int) (*SecretsManagerStore, error) {
+	session, region, err := getSession(numRetries)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := secretsmanager.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	stsSvc := sts.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	return &SecretsManagerStore{
+		svc:    svc,
+		stsSvc: stsSvc,
+	}, nil
+}
+
+// Write writes a given value to a secret identified by id.  If the secret
+// already exists, then write a new version.
+func (s *SecretsManagerStore) Write(id SecretId, value string) error {
+	version := 1
+	// first read to get the current version
+	latest, err := s.readLatest(id.Service)
+	mustCreate := false
+
+	if err != nil {
+		if len(value) == 0 {
+			return err
+		}
+		if err != ErrSecretNotFound {
```

Comment:
when `err == ErrSecretNotFound`, shouldn't `mustCreate` be set to `true`?

Reply:
Ah, no. This is a very tricky bit of code.

`ErrSecretNotFound` is the Chamber internal error, but `mustCreate` means that "we must create the underlying AWS Secrets Manager secret." Because this implementation doesn't have a 1:1 mapping of Chamber secret to AWS Secrets Manager secret, `ErrSecretNotFound` doesn't mean we need to create the underlying Secrets Manager secret. In fact, we will not encounter `ErrSecretNotFound` in this code path if the underlying AWS Secrets Manager secret needs to be created. If the AWS Secrets Manager secret does not exist, when we try to get it, we'll get the AWS error `secretsmanager.ErrCodeResourceNotFoundException` instead. That is the only case in which the underlying AWS Secrets Manager secret needs to be created.

Does this make sense? I'll add a code comment to explain this here.

Reply:
Added a comment via 44c774e